core/src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -2551,93 +2857,150 @@ deintercalate
             Partial _ _ -> error "Partial in extract"
-
-    extract (DeintercalateR fs sR consumed) = do
-        r <- extractR sR
-        case r of
-            Done n b -> processResult n fs (Right b)
             Error err -> return $ Error err
-            -- XXX Review consumed
-            Continue n s -> return $ Continue n (DeintercalateR fs s consumed)
+    extract (Deintercalate1InitR fs) = fmap (Done 0) $ fextract fs
+    extract (Deintercalate1R cnt fs _) = fmap (Done cnt) $ fextract fs
+    extract (Deintercalate1RL cnt bR fs sL) = do
+        r <- extractL sL
+        case r of
+            Done n bL -> do
+                res <- fstep fs (Right bR)
+                case res of
+                    FL.Partial fs1 -> extractResult n fs1 (Left bL)
+                    FL.Done _ -> error "Fold terminated consuming partial input"
+            Continue n s -> return $ Continue n (Deintercalate1RL (cnt - n) bR fs s)
             Partial _ _ -> error "Partial in extract"
+            Error _ -> do
+                xs <- fextract fs
+                return $ Done cnt xs
 
+{-# ANN type SepByState Fuse #-}
 data SepByState fs sp ss =
-      SepByInit !fs !sp
-    | SepBySeparator !fs !ss !Bool
+      SepByInitL !fs
+    | SepByL !Int !fs !sp
+    | SepByInitR !fs
+    | SepByR !Int !fs !ss
 
--- | Run the content parser first, when it is done, the separator parser is
--- run, when it is done content parser is run again and so on. If none of the
--- parsers consumes an input then parser returns a failure.
+-- | Apply two parsers alternately to an input stream. Parsing starts at the
+-- first parser and stops at the first parser. The output of the first parser
+-- is emiited and the output of the second parser is discarded. It can be used
+-- to parse a infix style pattern e.g. p1 p2 p1 . Empty input or single parse
+-- of the first parser is accepted.
 --
--- >>> sepBy p1 p2 sink = Parser.deintercalate p1 p2 (Fold.catLefts sink)
--- >>> sepBy content sep sink = Parser.sepBy1 content sep sink <|> return mempty
+-- Definitions:
+--
+-- >>> sepBy p1 p2 f = Parser.deintercalate p1 p2 (Fold.catLefts f)
+-- >>> sepBy p1 p2 f = Parser.sepBy1 p1 p2 f <|> Parser.fromEffect (Fold.extractM f)
+--
+-- Examples:
+--
+-- >>> p1 = Parser.takeWhile1 (not . (== '+')) Fold.toList
+-- >>> p2 = Parser.satisfy (== '+')
+-- >>> p = Parser.sepBy p1 p2 Fold.toList
+-- >>> Stream.parse p $ Stream.fromList ""
+-- Right []
+-- >>> Stream.parse p $ Stream.fromList "1"
+-- Right ["1"]
+-- >>> Stream.parse p $ Stream.fromList "1+"
+-- Right ["1"]
+-- >>> Stream.parse p $ Stream.fromList "1+2+3"
+-- Right ["1","2","3"]
 --
 {-# INLINE sepBy #-}
 sepBy :: Monad m =>
     Parser a m b -> Parser a m x -> Fold m b c -> Parser a m c
+-- This has similar performance as the custom impl below.
+-- sepBy p1 p2 f = deintercalate p1 p2 (FL.catLefts f)
 sepBy
-    (Parser pstep pinitial pextract)
-    (Parser sstep sinitial _)
+    (Parser stepL initialL extractL)
```

Comment:
This should be exactly the same as `deintercalate` except for the folding end resullts.

Reply:
States are simpler.  As the docs say:

-- >>> sepBy p1 p2 f = Parser.deintercalate p1 p2 (Fold.catLefts f)